spec/spec.md

```diff
@@ -0,0 +1,109 @@
+
+The spec will be worked on in the main branch and include the `-alpha` suffix until it reached a stable version. Once the first stable version of the spec is reached, it will be under the release-x.y branch and tagged with vX.Y.0, such as the runtime release process.
+
+This allows SDK developers to claim compliance towards a fixed spec and runtime version instead of a moving target.
+
+### SDK Versioning
+
+An SDK MUST start as one repository in the [dapr-sandbox organization](https://github.com/dapr-sandbox) at version 0.1, following the process to onboard a new Dapr sandbox project. Before an SDK can graduate to the [dapr organization](https://github.com/dapr), it MUST comply to a stable version of this spec, in addition to the graduation criteria for Dapr sandbox projects.
+
+## Client Object
+
+The SDK MUST offer a way to instantiate a client object without requiring any parameter, using the `DAPR_HTTP_PORT` or `DAPR_GRPC_PORT` environment variables injected by the Dapr sidecar injector or Dapr CLI - defaulting to `3500` or `50001`, correspondingly, if environment variable is not set.
+
+The client object MUST connect to the Dapr sidecar based on the `DAPR_SIDECAR_IP` environment variable, defaulting to `127.0.0.1` if not set.
+
+The client object MUST abstract the protocol used to communicate to the Dapr sidecar and not offer gRPC or HTTP specific input or output.
+
+The client object MIGHT offer both gRPC or HTTP implementations.
+
+The client object SHOULD NOT implement any Actor functionality since that is handled as a separate construct.
+
+The client object MUST implement all required methods for all [core building blocks](#core-building-blocks).
+
+The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.
+
+The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)
+
+## Object Serialization
+
+The SDK MUST serialize domain objects to JSON by default.
+
+The SDK MIGHT accept a serializer implementation via a provided interface.
+
+The SDK MIGHT accept a separate state serializer implementation via a provided interface to differentiate serialization of transient domain objects vs persisted domain objects.
+
+## Exception Handling
+
+Behavior and structure for exceptions is not defined yet.
+
+## Core Building Blocks
+
+* [State Store](core/state-store.md)
+* [PubSub](core/pubsub.md)
+* [Input Binding](core/input-binding.md)
+* [Output Binding](core/output-binding.md)
+* [Secrets Store](core/secrets-store.md)
+
+## Actors
+
+An SDK MIGHT implement the Actors building block.
+
+* [Actor client](actors/client.md)
+* [Actor runtime](actors/runtime.md)
+
+## Preview APIs
+
+The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:
```

Comment:
I'm not a huge fan of this decision. The way we handle this in the .NET SDK is by using the `[Obsolete]` attribute. This attribute forces the caller to acknowledge that the API is not stable (compilation failure if you don't). 

Using a preview client also means that you'll end up with a breaking change when an API graduates to stable as you'll need to update your code to use the standard client. This could be avoided by having the preview client map over to the standard client but then you just have additional code paths.

Reply:
This is an interesting discussion and it shows the need for a spec. I don't think we should optimize for no breaking changes for preview APIs graduating since they will break at the API level regardless. I am not sure if [obsolete] is giving the right message either since the API is not obsolete, it is "experimental".

I think the solution should make it clear that the app is using an Alpha API. Eager to see what others think here. We did this PreviewClient in the Java SDK and it looked pretty clean from the core client. It also makes it clear that it does not have the same guarantees as the core client.

Reply:
Yeah, I agree that `[Obsolete]` isn't the best attribute, but it's unfortunately the only one of that style in .NET. It does let you provide a message though, so when your code fails to compile it'll say something like "The ... API is current an Alpha API and is subject to change..."

The way we did it in the .NET SDK is that you won't need to have a breaking change when we change the API. The `DaprClient` is built as a layer over the gRPC provided client. So, when we update the proto, the methods will update. We can then call the new methods in the SDK and everything goes back to normal. When the user updates to the new revision, all they'll notice is that the `Obsolete` is gone but the compilation path remains the same. 

For languages that don't have support for this, it does become a bit harder. I wonder if we can do something else like an extension onto the client? You can import a separate package or project that contains those methods but they're labeled the same as the alpha APIs are an interface/inheritance/extension. When the graduate, they're moved to the base class and out of the alpha package. Thoughts?

Reply:
Interesting. If they are in a different package, then it would need to be a different interface, right? I am thinking of how to make the spec language agnostic. So, asking to have a different interface will accomplish that in a way that is more language neutral. In Java SDK, it is a separate interface. The class implements both interfaces but from the user's perspective, he is getting another type when we asks for an alpha client. I don't think we should optimize not to have a breaking change in SDK when an API graduates.

Reply:
> The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object.

My feedback is that this will make it too hard for users to use preview features in practice. In a complex business application, objects like a Dapr client are threaded through several layers of glue code to get to where they are actually called. This puts a lot of burden on the user to update these references to refer to the new **preview** type, and then update them again in the future to refer to the stable one.  This kind of idea can work well for dynamic type systems, but it's also less meaningful when you've got a dynamic type system.

Of the problems SDKs/Libraries face, there are two typical problems with preview features:

- Users are not aware of the preview features and thus, they don't get a lot of feedback until they are stable.
- Users accidentally use preview features and have mismatched expectations.

In my experience the former problem is way more common, so I wouldn't want to put up even more barriers.

I'd suggest being less prescriptive and talking about the goal instead. Something like:

> SDKs SHOULD use an in-code mechanism in addition to documentation to communicate to users about the instability of preview features. For a function to interface with the new **widget** building block could be called `Preview_GetWidget()`

Reply:
Since this prescription seems to be controversial. I will remove it and allow us to have a separate discussion about it.